cmd/collectors/rest/plugins/snapmirror/snapmirror.go

```diff
@@ -0,0 +1,188 @@
+ */
+package snapmirror
+
+import (
+	"encoding/json"
+	"github.com/tidwall/gjson"
+	"goharvest2/cmd/poller/plugin"
+	"goharvest2/cmd/tools/rest"
+	"goharvest2/pkg/conf"
+	"goharvest2/pkg/errors"
+	"goharvest2/pkg/matrix"
+	"strings"
+	"time"
+)
+
+const PluginInvocationRate = 10
+
+type SnapMirror struct {
+	*plugin.AbstractPlugin
+	client         *rest.Client
+	query          string
+	nodeUpdCounter int
+	svmVolToNode   map[string]string
+}
+
+func New(p *plugin.AbstractPlugin) plugin.Plugin {
+	return &SnapMirror{AbstractPlugin: p}
+}
+
+func (my *SnapMirror) Init() error {
+
+	var err error
+
+	if err = my.InitAbc(); err != nil {
+		return err
+	}
+
+	timeout := rest.DefaultTimeout * time.Second
+	if my.client, err = rest.New(conf.ZapiPoller(my.ParentParams), timeout); err != nil {
+		my.Logger.Error().Stack().Err(err).Msg("connecting")
+		return err
+	}
+
+	if err = my.client.Init(5); err != nil {
+		return err
+	}
+
+	my.query = "api/private/cli/volume"
+	my.svmVolToNode = make(map[string]string)
+
+	// Assigned the value to nodeUpdCounter so that plugin would be invoked first time to populate cache.
+	my.nodeUpdCounter = PluginInvocationRate
+
+	return nil
+}
+
+func (my *SnapMirror) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
+
+	if my.nodeUpdCounter >= PluginInvocationRate {
+		my.nodeUpdCounter = 0
+		if err := my.updateNodeCache(); err != nil {
+			return nil, err
+		}
+		my.Logger.Debug().Msg("updated node cache")
+	}
+
+	// update volume instance labels
+	my.updateSMLabels(data)
+	my.nodeUpdCounter++
+
+	return nil, nil
+}
+
+func (my *SnapMirror) updateNodeCache() error {
+	var (
+		records []interface{}
+		content []byte
+		err     error
+	)
+	href := rest.BuildHref("", "node", nil, "", "", "", "", my.query)
+
+	err = rest.FetchData(my.client, href, &records)
+	if err != nil {
+		my.Logger.Error().Stack().Err(err).Str("href", href).Msg("Failed to fetch data")
+		return err
+	}
+
+	all := rest.Pagination{
+		Records:    records,
+		NumRecords: len(records),
+	}
+
+	content, err = json.Marshal(all)
+	if err != nil {
+		my.Logger.Error().Err(err).Str("ApiPath", my.query).Msg("Unable to marshal rest pagination")
+	}
+
+	if !gjson.ValidBytes(content) {
+		my.Logger.Error().Err(err).Str("Api", my.query).Msg("Invalid json")
+		return errors.New(errors.API_RESPONSE, "Invalid json")
+	}
+
+	results := gjson.GetManyBytes(content, "num_records", "records")
+	numRecords := results[0]
+	if numRecords.Int() == 0 {
+		return errors.New(errors.ERR_NO_INSTANCE, "no "+my.query+" instances on cluster")
+	}
+
+	for _, volume := range results[1].Array() {
+		volumeName := volume.Get("volume").String()
+		vserverName := volume.Get("vserver").String()
+		nodeName := volume.Get("node").String()
+		my.svmVolToNode[vserverName+volumeName] = nodeName
+	}
+	return nil
+}
+
+func (my *SnapMirror) updateProtectedFields(instance *matrix.Instance) {
+
+	// check for group_type
+	if instance.GetLabel("group_type") != "" {
```

Comment:
this block looks pretty similar to a block in volume.go - can each plugin share this same block?

Reply:
it's almost same, moved to one util class and re-used.